ipa-core/src/protocol/ipa_prf/prf_sharding/mod.rs

```diff
@@ -306,27 +341,42 @@ pub trait GroupingKey {
     fn get_grouping_key(&self) -> u64;
 }
 
-#[tracing::instrument(name = "compute_histogram_of_users_with_row_count", skip_all)]
-pub fn compute_histogram_of_users_with_row_count<S>(input: &[S]) -> Vec<usize>
+#[tracing::instrument(name = "compute_histogram_range_sort_key_per_user", skip_all)]
+pub fn compute_histogram_range_sort_key_per_user<BK, TV, TS>(
+    input: &mut [PrfShardedIpaInputRow<BK, TV, TS>],
+) -> (Vec<usize>, Vec<Range<usize>>)
 where
-    S: GroupingKey,
+    BK: SharedValue,
+    TV: SharedValue,
+    TS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
 {
     let mut histogram = vec![];
     let mut last_prf = input[0].get_grouping_key() + 1;
     let mut cur_count = 0;
-    for row in input {
+    let mut start = 0;
+    let mut ranges = vec![];
+    for (idx, row) in input.iter_mut().enumerate() {
         if row.get_grouping_key() == last_prf {
             cur_count += 1;
         } else {
+            // Don't push upto single rows of users
```

Comment:
It's fine to do so as they'll just get filtered out. I would just remove this to keep the logic simpler.